### PR TITLE
feat(workflow): %-based Issue Rule Condition

### PR DESCRIPTION
**Pull request by iProgramStuff**
This PR aims to add a new condition to issue rules that can alert you when an issue has more errors than X% of sessions received over Y time period.

This builds on the `BaseEventFrequencyCondition ` class, counting the issue in the same way as `EventFrequencyCondition`, getting the count of sessions in the same time period, and dividing them.

Right now the interval is set to fixed increments that are 1 hour or less so that we can get data from the sessions table without needing to perform joins between the raw + hourly rollup tables. 

The tests have been modified so that they don't rely on snubas caching to be disabled by splitting up the tests per time period so that we have one that adds events and passes, and one that doesn't add events and fails.

### CODE REVIEW

File: src/sentry/rules/conditions/event_frequency.py
Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py:136]: Discussed with Adhi, we're going to just launch with the 1 hour lookback and look into the hybrid approach in the future

File: tests/snuba/rules/conditions/test_event_frequency.py
Comment (wedamija) [inline tests/snuba/rules/conditions/test_event_frequency.py:1]: Thanks for fixing this

File: src/sentry/rules/conditions/event_frequency.py
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:155]: Wondering if this should be `EventFrequencyPercentCondition` instead. That way, when we migrate this over to use `metrics` it doesn't need to be renamed.

File: src/sentry/rules/conditions/event_frequency.py
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:174]: Why do we fetch 5 days on each side?

File: src/sentry/rules/conditions/event_frequency.py
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:185]: Should we be summing all the values here per project?

File: src/sentry/constants.py
Comment (wedamija) [inline src/sentry/constants.py:225]: We should probably only display this behind a feature flag, at least initially.

Also wondering if we should only display this for orgs that have sessions

File: src/sentry/rules/conditions/event_frequency.py
Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py:174]: Oops, I added this in when testing something and never removed it :|

File: src/sentry/constants.py
Comment (iProgramStuff) [reply src/sentry/constants.py:225]: Yeah you're right. I'll figure out how we want to feature-gate this and add that in

Comment (wedamija) [reply src/sentry/constants.py:225]: I vaguely remember inter jeff did this by adding in an if statement into the endpoints that return the list of Conditions. I feel like @davidenwang might have also done something similar

Comment (davidenwang) [reply src/sentry/constants.py:225]: yup I think the easiest way to do this is add an if statement into the `ProjectRulesConfigurationEndpoint` to check for the feature flag and skip over adding the condition to the Response.

File: src/sentry/rules/conditions/event_frequency.py
Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py:185]: Nah, I don't think so because we are filtering by a single project. I was doing some testing recently and noticed that this will have to be summed up over time if/when we're querying for ranges greater than an hour (for now, we can keep the rollup at 60 and everything is in one bucket, but if rollup becomes smaller than the time range of the query, we see results like:
```[{'project_id': 2, 'bucketed_started': '2021-05-31T16:37:00+00:00', 'sessions': 2}, {'project_id': 2, 'bucketed_started': '2021-05-31T16:35:00+00:00', 'sessions': 2}]```.

File: src/sentry/constants.py
Comment (iProgramStuff) [reply src/sentry/constants.py]: Thanks for the tip. Implemented it that way and created the flag `organizations:issue-percent-filters`

File: src/sentry/rules/conditions/event_frequency.py
Comment (evanh) [inline src/sentry/rules/conditions/event_frequency.py:179]: I don't think you need to group by project_id here at all.

File: src/sentry/rules/conditions/event_frequency.py
Comment (evanh) [inline src/sentry/rules/conditions/event_frequency.py:164]: Sorry I'm new to this code, so I'm just trying to clarify:

This function gets run for every event that comes in. However you're caching the sessions results for one hour. Is start and end always 1 minute apart?

Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py:164]: Hey Evan. I'm just going to cache the session query for 5 minutes, that was the original decision that felt reasonable. The distance between start and end depend on the interval the user chooses for the rule and can be 1m, 5m, 10m, 30m, or 1h.

Comment (evanh) [reply src/sentry/rules/conditions/event_frequency.py:164]: OK that makes sense. For the record, given a time interval of 1m, the cache will refresh in less than 5 minutes as the interval straddles the hour. I don't think that's an issue though.

File: src/sentry/rules/conditions/event_frequency.py
Comment (evanh) [inline src/sentry/rules/conditions/event_frequency.py:170]: If `environment_id` is optional, shouldn't `project_id` or event ID be part of the cache_key? Otherwise every event from all projects with an empty `environment_id` will be reading from the same cached data.

File: src/sentry/rules/conditions/event_frequency.py
Comment (evanh) [inline src/sentry/rules/conditions/event_frequency.py:174]: I'm a little confused by this rollup as well: if there is an interval of 1 hour, then you are saying to group sessions by `toStartOfMinute(bucketed_started)`, which means you'll get up to 60 results back. You're then only selecting the first value below. If you only need one minute worth of data, why query for all 60?

File: src/sentry/rules/conditions/event_frequency.py
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:172]: I feel like we should actually be querying the last hour of data and then dividing this value. So if we have a 1 minute rule, query the last hour, then / 60. If we have a 5 minute rule, / 12, etc. This should help us avoid variability from sessions coming in late.

Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py:172]: Oof, I wrote a response to this yesterday but apparently didn't post it and now it's lost.

This is an interesting idea that I could see actually helping. It could potentially miss spikes though? It seems like it'd have the effect of firing an alert too early (if sessions spike the calculation will use a lower session value than it would today, making the % higher than maybe it should be), and the other side is if sessions drop we might actually "miss" an alert (the % would be lower due to the higher session count averaged over the last hour).

In practice, I don't expect sessions to fluctuate in the same way errors do. I'm happy to try this...we can maybe try less time too (30 minutes). What do you think of following up with this while we're able to poke around as-is with the condition in production?

File: src/sentry/rules/conditions/event_frequency.py
Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py:170]: Oops! Yeah this is a great catch. Thanks Evan.

File: src/sentry/rules/conditions/event_frequency.py
Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py:164]: Yeah definitely something good to be aware of, but I agree that it's not really an issue. Though, I am now feeling like I should almost write a comment about that edge case or just fix it somehow.

File: src/sentry/rules/conditions/event_frequency.py
Comment (wedamija) [reply src/sentry/rules/conditions/event_frequency.py:172]: The main problem here is that sessions arrive a little late, so fetching sessions for the last minute is unlikely to be accurate. I think at the very least, you'd need to fetch them from the last 5-10 minutes.

Since we're just testing this for now, I'm ok with leaving it as is for the moment.

File: src/sentry/rules/conditions/event_frequency.py
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:164]: Now that I look at this cache key, I'm not actually sure if it's right. Since start/end can be for different time ranges (1 minute, 5 minutes, etc), aren't the different filters going to end up stomping on each other because we drop off the minute part here?

My initial thought here was that we'd be fetching this for the last hour and caching that, and using the same value across all filters. If we want to handle it this way, then we need to make sure we vary the cache by filter type

File: src/sentry/rules/conditions/event_frequency.py
Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py:175]: Hmm, so I made the rollup 60 so that we'd hit the raw table. When I did that, I definitely needed to start summing the results in every bucket (at the time of your comment, it would have only been getting correct counts for 1m intervals). Just fixed that. Previously there was no bucketing and only one result of the session count, so I didn't sum them up.

It sounds like you would prefer I use a rollup of 3600. If I make the rollup 3600, snuba groups using `toStartOfHour` which works, but it also starts hitting the hourly table.

So if I need one minute of data from the raw table, right now the only way I know of to do this is to make the rollup 60 and use start/end to limit the query. If start/end are 1 minute apart, I only seem to get one bucket back. If they are 10 minutes apart, I seem to get max ten back (if there were sessions every minute). This seems basically fine to me.

So now I sum up the results, and I modified the test to better spread the sessions throughout the time period so that without summing it will actually fail. What do you think?

File: src/sentry/rules/conditions/event_frequency.py
Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py]: Aaah you're completely right. Your idea gets around that problem well too. I thought the delay was built in to the snuba side but I think I was wrong about that. Caching on interval here works better I think and resolves the edge case of when start/end are on the boundary of an hour. Gonna do that.

File: src/sentry/rules/conditions/event_frequency.py
Comment (evanh) [reply src/sentry/rules/conditions/event_frequency.py:175]: Now that you're using all the results it's fine. I only mentioned a limit because you were only using 1 row of data even though you were getting back more than that. The 60 second rollup is fine.

File: src/sentry/rules/conditions/event_frequency.py
Comment (evanh) [inline src/sentry/rules/conditions/event_frequency.py:165]: You didn't actually change this to use the interval. If you take Dan's idea of querying for the last hour of data and dividing, it might save snuba queries as well because there wouldn't be a separate cache for each time interval. Then in your percentage calculation below you would have to account for the fact that session_count is for a full hour.

File: src/sentry/rules/conditions/event_frequency.py
Comment (evanh) [inline src/sentry/rules/conditions/event_frequency.py:191]: If session_count is 0, but issue_count > 1, isn't that 100%? Or do you default to 0 in the case of the user not having the correct data?

Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py:191]: I think we're going to default to 0 here so that the alert doesn't fire in this case, at least for now.

File: src/sentry/rules/conditions/event_frequency.py
Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py]: Just pushed up an update using this idea. I was thinking of storing `end` without minutes, but I think it's okay since this function is only ever called with end basically always being the current time.

So now I'm averaging out the last hour of data and saving the extra snuba queries.

File: src/sentry/rules/conditions/event_frequency.py
Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py]: Implemented!

File: tests/sentry/api/endpoints/test_project_agnostic_rule_conditions.py
Comment (wedamija) [inline tests/sentry/api/endpoints/test_project_agnostic_rule_conditions.py:14]: Should this be returning 10 items if the new feature flag is working?

File: src/sentry/rules/conditions/event_frequency.py
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:165]: Why do we need to remove spaces here? These should both just be ints, I think?

File: src/sentry/rules/conditions/event_frequency.py
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:184]: `sum(bucket["sessions"] for bucket in result_totals["data"])`

File: src/sentry/rules/conditions/event_frequency.py
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:192]: Do you want this to be py2 style division? ie `//`? That way you get whole minutes

File: src/sentry/rules/conditions/event_frequency.py
Comment (wedamija) [inline src/sentry/rules/conditions/event_frequency.py:202]: I think @adhiraj mentioned this in slack, but it probably does make sense to make sure `issue_count < avg_sessions_in_interval`. Doesn't have to be in this pr

File: tests/snuba/rules/conditions/test_event_frequency.py
Comment (wedamija) [inline tests/snuba/rules/conditions/test_event_frequency.py:157]: These tests take quite a while to run, 45-50 seconds... wondering if we need to create 60 session buckets for all the tests?

Alternatively, I'm wondering if we can just create sessions/events in bulk. `store_session` actually puts the session into a list, so we could probably have an alternate version that just accepts all the sessions at once.

**Comment by evanh** [comment]
This looks good to me but I'll let Dan approve after you address his comments.

File: tests/sentry/api/endpoints/test_project_agnostic_rule_conditions.py
Comment (iProgramStuff) [reply tests/sentry/api/endpoints/test_project_agnostic_rule_conditions.py:14]: Oh, missed this! Good catch. I only modified `project_rules_configuration.py`

File: src/sentry/rules/conditions/event_frequency.py
Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py:165]: Removing it now. Left it in carelessly from when I had the time in here

File: tests/snuba/rules/conditions/test_event_frequency.py
Comment (iProgramStuff) [reply tests/snuba/rules/conditions/test_event_frequency.py:157]: Yeah I was actually thinking these tests were taking awhile. Part of that problem was introduced when we queried over the hour. I an making 60 sessions right now so that I get 1 session per minute on average and can make about half the events. Can speed this up a bit by making less sessions and customizing the tests a bit more and try that bulk method

File: src/sentry/rules/conditions/event_frequency.py
Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py:192]: Yeah, I think that seems like a safer bet. The interval is always whole in tests right now. What do you think of pulling it from the rule option via `percent_intervals[self.get_option("interval")][1].total_seconds() / 60`?

File: tests/snuba/rules/conditions/test_event_frequency.py
Comment (iProgramStuff) [reply tests/snuba/rules/conditions/test_event_frequency.py:157]: Running just `EventFrequencyPercentConditionTestCase` locally, I see the tests are about twice as fast by just implementing a bulk method (from about ~30->~15 seconds on average for me)

File: src/sentry/rules/conditions/event_frequency.py
Comment (wedamija) [reply src/sentry/rules/conditions/event_frequency.py]: I think that'd be safer tbh, any reason that we shouldn't?

File: tests/snuba/rules/conditions/test_event_frequency.py
Comment (wedamija) [reply tests/snuba/rules/conditions/test_event_frequency.py:157]: Nice, it'd be good to introduce a bulk event creation method too, but I think that would take a bit more work.

File: src/sentry/rules/conditions/event_frequency.py
Comment (iProgramStuff) [reply src/sentry/rules/conditions/event_frequency.py]: I think it seems okay to pull it from the rule option. I can't think of a situation where they should be different anyways. Updated

**Review by wedamija — APPROVED** [review]

**Pull request merged**
